From 564ba91e8e91fb7fefc1772d937bcd126d0ff84b Mon Sep 17 00:00:00 2001
From: Dennis Ostermann <dennis.ostermann@renesass.com>
Date: Wed, 7 Dec 2022 14:34:22 +0100
Subject: [PATCH 131/151] RSwitch2: Minor fixes

- Do RX dma_unmap() on exit
- If no correct speed is set, use default speed for SerDes init
---
 .../ethernet/renesas/rswitch2/rswitch2_eth.c  | 66 ++++++++++++-------
 1 file changed, 41 insertions(+), 25 deletions(-)

diff --git a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
index 7dfeb2e45977..655af4f1fb8f 100644
--- a/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
+++ b/drivers/net/ethernet/renesas/rswitch2/rswitch2_eth.c
@@ -197,7 +197,7 @@ static int rswitch2_serdes_reg_wait(void __iomem *addr, u32 offs, u32 bank, u32
 	return -ETIMEDOUT;
 }
 
-static int rswitch2_serdes_set_speed(struct rswitch2_eth_port *eth_port)
+static int rswitch2_serdes_set_speed(struct rswitch2_eth_port *eth_port, int phy_speed)
 {
 	struct rswitch2_drv *rsw2;
 	struct rswitch2_physical_port *phy_port;
@@ -206,15 +206,15 @@ static int rswitch2_serdes_set_speed(struct rswitch2_eth_port *eth_port)
 	phy_port = eth_port->phy_port;
 	rsw2 = eth_port->rsw2;
 
-	rsw2_notice(MSG_SERDES, "Set serdes speed: %d\n", phy_port->phy->speed);
+	rsw2_notice(MSG_SERDES, "Set serdes speed: %d\n", phy_speed);
 
 	switch (phy_port->phy_iface) {
 	case PHY_INTERFACE_MODE_SGMII:
-		if (phy_port->phy->speed == SPEED_1000)
+		if (phy_speed == SPEED_1000)
 			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x0140);
-		else if (phy_port->phy->speed == SPEED_100)
+		else if (phy_speed == SPEED_100)
 			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x2100);
-		else if (phy_port->phy->speed == SPEED_10)
+		else if (phy_speed == SPEED_10)
 			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x0100);
 		else
 			return -EOPNOTSUPP;
@@ -228,7 +228,7 @@ static int rswitch2_serdes_set_speed(struct rswitch2_eth_port *eth_port)
 		break;
 
 	case PHY_INTERFACE_MODE_USXGMII:
-		if (phy_port->phy->speed == SPEED_2500)
+		if (phy_speed == SPEED_2500)
 			rswitch2_serdes_write32(addr, SR_MII_CTRL, BANK_1F00, 0x120);
 		else
 			return -EOPNOTSUPP;
@@ -768,6 +768,7 @@ static int rswitch2_serdes_init(struct net_device *ndev, bool check_op)
 	struct rswitch2_drv *rsw2;
 	struct rswitch2_eth_port *eth_port;
 	struct rswitch2_physical_port *phy_port;
+	int phy_speed;
 	uint port_num;
 
 	eth_port = netdev_priv(ndev);
@@ -787,13 +788,25 @@ static int rswitch2_serdes_init(struct net_device *ndev, bool check_op)
 	//printk("%s: sw0p%d  phy_mode=%d\n", __FUNCTION__, port_num, ndev->phydev->interface );
 	switch (phy_port->phy_iface) {
 	case PHY_INTERFACE_MODE_SGMII:
-		//phy_port->phy->speed = SPEED_1000;
-		rsw2_notice(MSG_SERDES, "port %d uses SGMII at %d Mbit/s\n", port_num, phy_port->phy->speed);
+		if((phy_port->phy->speed != SPEED_100) && (phy_port->phy->speed != SPEED_1000)) {
+			phy_speed = SPEED_100;
+			rsw2_notice(MSG_SERDES, "No valid default speed. Setting to %d Mbit/s\n", phy_speed);
+		} else
+			phy_speed = phy_port->phy->speed;
+
+		rsw2_notice(MSG_SERDES, "port %d uses SGMII at %d Mbit/s\n", port_num, phy_speed);
 		break;
+
 	case PHY_INTERFACE_MODE_USXGMII:
-		//phy_port->phy->speed = SPEED_2500;
-		rsw2_notice(MSG_SERDES, "port %d uses USXGMII at %d Mbit/s\n", port_num, phy_port->phy->speed);
+		if(phy_port->phy->speed != SPEED_2500) {
+			phy_speed = SPEED_2500;
+			rsw2_notice(MSG_SERDES, "No valid default speed. Setting to %d Mbit/s\n", phy_speed);
+		} else
+			phy_speed = phy_port->phy->speed
+			;
+		rsw2_notice(MSG_SERDES, "port %d uses USXGMII at %d Mbit/s\n", port_num, phy_speed);
 		break;
+
 	default:
 		rsw2_err(MSG_SERDES, "%s: Don't support this interface %d on port %d", __func__,
 				phy_port->phy_iface, port_num);
@@ -809,7 +822,7 @@ static int rswitch2_serdes_init(struct net_device *ndev, bool check_op)
 	}
 
 	/* Set speed (bps) */
-	ret = rswitch2_serdes_set_speed(eth_port);
+	ret = rswitch2_serdes_set_speed(eth_port, phy_speed);
 	if (ret) {
 		rsw2_err(MSG_SERDES, "set initial SERDES speed for port %d failed\n", port_num);
 		return ret;
@@ -996,15 +1009,7 @@ static int rswitch2_eth_open(struct net_device *ndev)
 		netif_tx_start_all_queues(ndev);
 
 
-
-
-
-
-		//ret = rswitch2_mac_set_speed(eth_port);
-		//if (ret != 0)
-		//	return ret;
-
-		ret = rswitch2_serdes_init(ndev, true);
+		ret = rswitch2_serdes_init(ndev, false);
 		if (ret != 0) {
 			rsw2_err(MSG_SERDES, "%s: rswitch2_serdes_init failed: %d\n", ndev->name, ret);
 			return ret;
@@ -2853,7 +2858,7 @@ static int rswitch2_mdio_read(struct mii_bus *bus, int addr, int regnum)
 	/* Not support Clause 22 access method */
 	if (!mode) {
 		// FIXME
-		//rsw2_warn(MSG_GEN,"NOT C45 regnum=%04x  return 0\n", regnum);
+		//printk("NOT C45 regnum=%04x  return 0\n", regnum);
 
 		return 0;
 	}
@@ -2872,7 +2877,7 @@ static int rswitch2_mdio_write(struct mii_bus *bus, int addr, int regnum, u16 va
 	/* Not support Clause 22 access method */
 	if (!mode) {
 		// FIXME
-		//rsw2_warn(MSG_GEN,"NOT C45 regnum=%04x  return 0\n", regnum);
+		//printk("NOT C45 regnum=%04x  return 0\n", regnum);
 
 		return 0;
 	}
@@ -3166,10 +3171,11 @@ static int rswitch2_init_physical_port(struct rswitch2_drv *rsw2, unsigned int p
 	netif_napi_add(port_ndev, &phy_port->rx_q[0].napi, rswitch2_poll, 64);
 
 	ret = phy_connect_direct(port_ndev, phy_port->phy, &rswitch2_phy_state_change, phy_port->phy_iface);
-
 	if (ret != 0) {
-		rsw2_err(MSG_GEN, "%s: rswitch2_phy_init failed: %d\n", port_ndev->name, ret);
+		rsw2_err(MSG_GEN, "%s: phy_connect_direct() failed: %d\n", port_ndev->name, ret);
 		return ret;
+	} else {
+		rsw2_notice(MSG_GEN, "phy_connect_direct(): attached '%s' PHY driver: %d\n", port_ndev->phydev->drv->name , ret);
 	}
 
 	rsw2->ports[rsw2->ports_intialized] = eth_port;
@@ -3282,6 +3288,12 @@ static void rswitch2_free_rings(struct net_device *ndev)
 	uint num_of_rx_queues;
 	uint num_of_tx_queues;
 	uint q;
+#ifdef RSW2_RX_TS_DESC
+	struct rswitch2_dma_ext_ts_desc *rx_desc;
+#else
+	struct rswitch2_dma_ext_desc *rx_desc;
+#endif
+
 
 	eth_port = netdev_priv(ndev);
 	rsw2 = eth_port->rsw2;
@@ -3308,8 +3320,12 @@ static void rswitch2_free_rings(struct net_device *ndev)
 		ring_size = sizeof(*rx_q->desc_ring) * (rx_q->entries + 1);
 
 		for(q_entry = 0; q_entry < rx_q->entries; q_entry++) {
-			if(rx_q->skb[q_entry] != NULL)
+			if(rx_q->skb[q_entry] != NULL) {
+				rx_desc = &rx_q->desc_ring[q_entry];
+				dma_unmap_single(ndev->dev.parent, le32_to_cpu(rx_desc->dptrl),
+									RSW2_PKT_BUF_SZ, DMA_FROM_DEVICE);
 				dev_kfree_skb_any(rx_q->skb[q_entry]);
+			}
 		}
 
 
-- 
2.17.1

